core/provisioning-java/src/main/java/org/apache/syncope/core/provisioning/java/pushpull/LDAPMembershipPullActions.java

```diff
@@ -155,8 +156,10 @@ public void beforeUpdate(
         groupDAO.findUMemberships(groupDAO.find(entity.getKey())).forEach(uMembership -> {
             Set<String> memb = membershipsBefore.get(uMembership.getLeftEnd().getKey());
             if (memb == null) {
-                memb = new HashSet<>();
-                membershipsBefore.put(uMembership.getLeftEnd().getKey(), memb);
+                memb = Collections.synchronizedSet(new HashSet<>());
```

Comment:
This snippet should be replaced by using `ConcurrentHashMap#computeIfAbsent`, which involves declaring both maps as `ConcurrentMap` - see https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentMap.html#computeIfAbsent-K-java.util.function.Function-

e.g.

```java
            Set<String> memb = membershipsBefore.computeIfAbsent(
                    uMembership.getLeftEnd().getKey(),
                    k -> Collections.synchronizedSet(new HashSet<>()));
            memb.add(entity.getKey());
```

Reply:
Thank your review, let me learned new API